update CentOS recipe to use CMAKE 3.5 and Qt5.6

ericfont opened the pull request:
@shoogle, I think we need to update the CentOS AppImage Recipe to use CMAKE 3.5 and Qt 5.6 (need to use QWebEngine instead of QWebKit).  I got CMAKE working here by using the same binaries that are used in test job.  But then I found that QWebEngine is not available in the CentOS 6 repos (and even in https://dl.fedoraproject.org/pub/epel/6/x86_64/repoview/letter_q.group.html I don't see any QWebEngine, just QWebKit).  So I tried just copying the same binaries for qt5.6 that are used in the test job, and pointed LD_LIBRARY_PATH there.  I don't know if that is appropriate.  But that just gives me a bunch of errors about not being able to find qt components (https://travis-ci.org/ericfont/MuseScore/jobs/123194377#L2534-L2641).  So I'm not sure what needs to be done now to get qt5.6 w/ QWebEngine working with the AppImage build.

ericfont commented:
fedora says QtWebEngine will be a new package, different from QWebKit.  https://fedoraproject.org/wiki/Changes/QtWebEngine

Now how to properly get this QWebEngine inside the centos 6 is the question.

ericfont commented:
maybe could disable webengine from CMAKE...there is already QT_USE_QTWEBKIT in CMakeLists.txt, which could be set to off...maybe need to update that switch to be for QTWEBENGINE and then have easy way to set that to OFF when call Makefile.  Do you now if we wait long enough, will EPEL update to QWebEngine?

lasconic commented:
Really, I wouldn't bother with the repo. Similar to Cmake, you can install Qt5.6 like this https://github.com/musescore/MuseScore/blob/master/build/travis/job1_Tests/environment.sh#L23

lasconic commented:
And now I read the code, and see you did it. 
However this is a problem with genManual
I had problems on mac and windows too but for different reason. It seems genManual is run and executed during the build phase, so it needs to be able to find the Qt libraries at runtime. They need to be in RPATH or LD_LIBRARY_PATH

shoogle commented:
I'm more than happy for you to make changes to the AppImages as you see fit. Maybe tag me in the PR so I stay in-the-loop (like you did here).

I didn't realise MuseScore's zip file of Qt included the Linux files (at least the x86_64 ones). It should be possible to get i686 binaries straight from http://www.qt.io/, but I think we'll probably have to wait for a repo version of Qt for ARM, or compile it ourselves.

I notice the AppImage builds are "passing" on Travis, even though they actually exit early with errors. This shouldn't happen - if there are errors they should fail. (The overall Travis build shouldn't fail though, unless the tests on Ubuntu fail).  But I can't see why this is happening since both build.sh and Recipe have `set -e`.

ericfont commented:
@shoogle Will "set -e" propagate an error condition from inside a Docker shell to the outside?

shoogle commented:
The Docker shell is supposed to exit with the status of the last command executed inside it, so if the recipe exits with an error then Docker should too.

ericfont commented:
well we have " || true" at the end of each AppImage build command in the travis.yml: https://github.com/musescore/MuseScore/blob/master/.travis.yml#L69

I've checked in my terminal and found that that " || true" will cause that to always report success (return 0):<pre> [[ 1 == 0 ]] || true && echo $?  #results in 0
[[ 1 == 1 ]] || true && echo $?  #results in 0</pre>
I can't remember the reason we put that " || true" at the end...I'm remember doing that so would always report successes if the APPIMAGE_BUILD_ARCHS != specific_arch.  But we don't want it to always report success.  We actually do want to report failure if the build fails.  So I think need to modify into if-then-else.

ericfont commented:
@shoogle Here was the discussion on that https://github.com/musescore/MuseScore/pull/2491#discussion_r57520469

I went with your first option, but I think I need to go with your 3rd option:
<pre>if [[ "$APPIMAGE_BUILD_ARCHS" == *armhf* ]]; then
  ./build
fi</pre>

Let me try that in travis and see if fails...

ericfont commented:
@shoogle I've changed the travis yml commands to "if; then; fi" format, and that seems to do the trick by reporting failure properly https://travis-ci.org/ericfont/MuseScore/jobs/123299348

ericfont commented:
ok, I think I fixed it by yum installing a few different libs `fontconfig libxslt libXcomposite libXcursor libXrender` which are needed by genManual.  They must have been dependencies of the CentOS Qt packages that we're not using anymore.

FYI, I think I'd like to make a special docker file for i686 CentOS 6 and x86_64 CentOS 6, which would probably shave off a couple minutes off the AppImage build.  I might also split the CentOS Recipe into a few different functions like I did in RecipeDebian:<pre>      fetch-build-dependencies
      fetch-package-dependencies
      build "$@"
      package
</pre>The reason being then I can easily build the CentOS 6 dockerfile by simply using command `RUN ./RecipeDebian --fetch-build-dependencies-only armhf` just like I did in jessie-crosscompile-armhf.Dockerfile

ericfont commented on build/Linux+BSD/portable/copy-libs:
```diff
@@ -79,8 +79,9 @@ getCrossPlatformDependencies() {
-  copyLib libQt5WebKitWidgets.so.5
+  copyLib libQt5WebEngine.so.5
+  copyLib libQt5WebEngineCore.so.5
+  copyLib libQt5WebEngineWidgets.so.5
```
The check-depends https://travis-ci.org/ericfont/MuseScore/jobs/123330905#L4731-L4744 reports:<pre># Provided by neither: 13
libcups.so.2
libEGL.so.1
libicudata.so.56
libicui18n.so.56
libicuuc.so.56
libQt5Bluetooth.so.5
libQt5Nfc.so.5
libQt5QuickParticles.so.5
libQt5WebChannel.so.5
libQt5WebEngine.so.5
libQt5WebEngineCore.so.5
libQt5WebEngineWidgets.so.5
libQt5WebSockets.so.5</pre>
Now I've just added WebEngine, Core, & Widgets, and I've updated the version numbers for icu, so those errors will go away on next build.  And we've already decided libQt5QuickParticles is not needed.  But what about these other libraries?  I'm guessing libQt5Bluetooth.so.5 and libQt5Nfc.so.5 won't be needed.  But what about libQt5WebSockets?

shoogle replied:
Your guess is as good as mine. @lasconic may have a better idea of what's needed.
Also see changes to Windows dependencies in [commit 4875862](https://github.com/musescore/MuseScore/commit/4875862aa19abdf204609b8e679bba3cbc785e0c).

lasconic replied:
I can't comment on the non Qt libs. For the Qt libs, I think only the 3 webengine one are needed by MuseScore binary but then why would check-depends report them?

shoogle replied:
They would get reported if they are linked to by one of the Qt libraries that is already included in the AppImage. For example:
- mscore depends on libA.so for feature X
- libA.so depends on libB.so for feature Y (but not for feature X)

In this case libB.so will be reported as missing by check-depends even though it is not actually needed by MuseScore. Probably one of the general Qt libs (e.g. libQt5Core.so) links to lots of other Qt libs.

Usually indirect dependencies will be needed, hence why they are reported as missing by check-depends. I'm planning on making an ignore list to avoid reporting ones we know we don't need.

ericfont commented on build/Linux+BSD/portable/copy-libs:
```diff
@@ -274,6 +275,11 @@ building_on_CentOS_6() {
+
+  # after upgrade to Qt 5.6, check-depends in CentOS 6 said following were provided by neither
+  copyLib libEGL.so.1
+  copyLib libcups.so.2
```
I've added EGL & cups, since were provided by neither in check-depends

ericfont commented:
Well looks like the x86-64 AppImage has now been successfully built in travis and I'm able to run it on my ArchLinux x86-64.

shoogle commented:
Nice job! I don't have time to look into fetching/building Qt for i686 or ARM, but feel free to have a look yourself if you do have time. If not, it's not urgent right now anyway (no releases planned) so we can just disable ARM and i686 AppImage for the time being.

**Goals:**
Short term - borrow x86_64 Qt libraries (merge this PR)
Mid term - build Qt ourselves for every architecture (or find a reliable source for every arch).
Long term - investigate static linking.

ericfont commented on build/Linux+BSD/portable/copy-libs:
```diff
@@ -259,9 +260,9 @@ building_on_CentOS_6() {
-  copyLib libicuuc.so.42 # moved from cross-platform
+  copyLib libicudata.so.56 # moved from cross-platform
+  copyLib libicui18n.so.56 # moved from cross-platform
+  copyLib libicuuc.so.56 # moved from cross-platform
```
I had to update this to .56 or else reported this missing check-depends.  This .56 is provided by the qt5 zipfile, while .42 is provided by CentOS 6 repo.

ericfont commented:
shoogle, sounds good...someother day I'll update i686 and ARM to qt5.6.  

I did have a working x86-64 build https://github.com/ericfont/MuseScore/commit/03bb51c56fdf0aafdefb98e6def613235d8600fe which produces an AppImage https://bintray.com/ericfont/MuseScoreDevelopment/MuseScoreDev-AppImageRecipeCMAKE-x86_64/201604151350-AppImageRecipeCMAKE-03bb51c which I'm able to run on my ArchLinux x86-64 Machine.  But then I made some changes based on output of check-depends, trying to improve, but unfortunately now my current commit is not producing an AppImage that will run on my computer.  I'm going to take a break now...so wait till I fix that before merge.

probonopd commented on build/Linux+BSD/portable/Recipe:
```diff
@@ -74,9 +74,23 @@ cd AppImageKit
+export PATH=${PWD}/cmake/bin:${PATH}
+
+# get Qt5.6 (Centos 6 doesn't have QtWebEngine)
+mkdir qt5 && wget -q -O qt5.zip http://utils.musescore.org.s3.amazonaws.com/qt560.zip
```
Where is qt560.zip coming from and how was it made? Might be of interest to other Qt-based [AppImages](https://github.com/probonopd/AppImages), too.

lasconic replied:
I installed Qt on a VM running a desktop and zipped the qt directory. It
would be cool if other projects host it and don't use our bandwidth.

lasconic replied:
An alternative is to download the official package and script it see https://github.com/benlau/qtci

lasconic commented:
I'm trying to revive this PR. See https://github.com/musescore/MuseScore/pull/2763

Jojo-Schmitz commented:
Can't this PR get closed then?

lasconic commented:
Done

lasconic commented:
If anyone on linux could check if the app image is working that would be helpful. https://bintray.com/musescore/nightlies-linux/MuseScoreNightly-master-x86_64/_latestVersion

shoogle commented:
It works on my Ubuntu, but not on a clean CentOS. I'll have to make some changes to [copy-libs](https://github.com/lasconic/MuseScore/blob/cca140e7beffd2d1b13cc871d74a151917ea2167/build/Linux%2BBSD/portable/copy-libs).

lasconic commented:
On IRC @hpfmn mentioned that it works on Arch and on the forum kuwitt said [it works with  OpenSuse Leap 42.1](https://musescore.org/en/node/116886#comment-543296)

lasconic commented:
On IRC, @MarcSabatella reported that it works ok on Ubuntu 14.04.
